crates/core/tests/large_network.rs

```diff
@@ -0,0 +1,327 @@
+        .ok()
+        .and_then(|val| val.parse().ok())
+        .unwrap_or(DEFAULT_PEER_COUNT);
+    let snapshot_interval = env::var("SOAK_SNAPSHOT_INTERVAL_SECS")
+        .ok()
+        .and_then(|val| val.parse().ok())
+        .map(Duration::from_secs)
+        .unwrap_or(DEFAULT_SNAPSHOT_INTERVAL);
+    let snapshot_iterations = env::var("SOAK_SNAPSHOT_ITERATIONS")
+        .ok()
+        .and_then(|val| val.parse().ok())
+        .unwrap_or(DEFAULT_SNAPSHOT_ITERATIONS);
+    let connectivity_target = env::var("SOAK_CONNECTIVITY_TARGET")
+        .ok()
+        .and_then(|val| val.parse::<f64>().ok())
+        .unwrap_or(DEFAULT_CONNECTIVITY_TARGET);
+    let snapshot_warmup = env::var("SOAK_SNAPSHOT_WARMUP_SECS")
+        .ok()
+        .and_then(|val| val.parse().ok())
+        .map(Duration::from_secs)
+        .unwrap_or(DEFAULT_SNAPSHOT_WARMUP);
+    let min_connections = env::var("SOAK_MIN_CONNECTIONS")
+        .ok()
+        .and_then(|val| val.parse().ok())
+        .unwrap_or(DEFAULT_MIN_CONNECTIONS);
+    let max_connections = env::var("SOAK_MAX_CONNECTIONS")
+        .ok()
+        .and_then(|val| val.parse().ok())
+        .unwrap_or(DEFAULT_MAX_CONNECTIONS);
+
+    let network = TestNetwork::builder()
+        .gateways(2)
+        .peers(peer_count)
+        .min_connections(min_connections)
+        .max_connections(max_connections)
+        .require_connectivity(connectivity_target)
+        .connectivity_timeout(Duration::from_secs(120))
+        .preserve_temp_dirs_on_failure(true)
+        .preserve_temp_dirs_on_success(true)
+        .binary(FreenetBinary::CurrentCrate(BuildProfile::Debug))
+        .build()
+        .await
+        .context("failed to start soak test network")?;
+
+    println!(
+        "Started soak network with {} gateways and {} peers (run root: {})",
+        2,
+        peer_count,
+        network.run_root().display()
+    );
+    println!(
+        "Min connections: {}, max connections: {} (override via SOAK_MIN_CONNECTIONS / SOAK_MAX_CONNECTIONS)",
+        min_connections, max_connections
+    );
+
+    let riverctl_path = which("riverctl")
+        .context("riverctl not found in PATH; install via `cargo install riverctl`")?;
+
+    let alice_url = format!("{}?encodingProtocol=native", network.gateway(0).ws_url());
+    let bob_url = format!("{}?encodingProtocol=native", network.peer(0).ws_url());
+    let session = RiverSession::initialize(riverctl_path, alice_url, bob_url).await?;
+
+    let snapshots_dir = network.run_root().join("large-soak");
+    fs::create_dir_all(&snapshots_dir)?;
+
+    // Allow topology maintenance to run before the first snapshot.
+    println!(
+        "Waiting {:?} before first snapshot to allow topology maintenance to converge",
+        snapshot_warmup
+    );
+    sleep(snapshot_warmup).await;
+
+    let mut iteration = 0usize;
+    let mut next_tick = Instant::now();
+    while iteration < snapshot_iterations {
```

Comment:
note for the future: I think we can do better than this... we already programatically collect events and have an event aggregator, no need to do this here, this can be moved to the event collector for tests.